### PR TITLE
Add self-contained artifact upload script for apache nexus (#11329)

### PR DESCRIPTION
**Pull request by dweiss**

### CODE REVIEW

**Comment by dweiss** [comment]
I initially wanted to use Python but without additional libraries it overwhelmed me and I wanted to keep it simple and self-contained. There is some extra verbosity (XML processing in Java is hell) but it does work and I staged artifacts for Lucene 10.0.0 successfully (and dropped the repository afterwards).

Now... I'm not so sure how to test the wizard bit. I based the changes on @HoustonPutman 's work on Solr side. The script can accept user name and password in environment variables, on command line or will prompt for password if not provided... but I haven't tested this as I'm not sure how to test it in isolation (@janhoy - is there any way to do it?).

**Comment by dweiss** [comment]
For stand-alone use, it's simple enough too:
```
java dev-tools\scripts\StageArtifacts.java -u dweiss /release/candidate/maven-artifacts
```
will prompt for password for nexus access.

**Comment by jpountz** [comment]
Thanks @dweiss I'll give it a try!

File: dev-tools/scripts/StageArtifacts.java [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+        var params = new Params();
+        for (int i = 0; i < args.length; i++) {
+          switch (args[i]) {
+            case "-n":
+            case "--nexus":
+              params.nexusUri = URI.create(args[++i]);
+              break;
+            case "-u":
+            case "--user":
+              params.userName = args[++i];
```
Comment (madrob) [inline dev-tools/scripts/StageArtifacts.java:86]: Can we create a separate method for getting the next arg that would throw something nicer than an IOOBException when somebody (i.e. me) inevitably goofs up?

File: dev-tools/scripts/StageArtifacts.java [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+              System.out.println("  -u, --user  User name for authentication.");
+              System.out.println("              better: ASF_USERNAME env. var.");
+              System.out.println("  -p, --password  Password for authentication.");
+              System.out.println("              better: ASF_PASSWORD env. var.");
+              System.out.println("  -n, --nexus URL to Apache Nexus (optional).");
+              System.out.println("  --description  Staging repo description (optional).");
+              System.out.println("");
+              System.out.println("  path        Path to maven artifact directory.");
+              System.out.println("");
+              System.out.println(" Password can be omitted for console prompt-input.");
```
Comment (madrob) [inline dev-tools/scripts/StageArtifacts.java:108]: Use multi-line string?

File: dev-tools/scripts/StageArtifacts.java [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+            null);
+
+        HttpResponse.BodyHandler<String> bodyHandler = HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8);
+        HttpRequest req = HttpRequest.newBuilder()
+            .GET()
+            .uri(target)
+            // we could use json if XML is too difficult to work with.
+            // .header("Accept", "application/json")
+            .build();
+        HttpResponse<String> response = client.send(req, bodyHandler);
```
Comment (madrob) [inline dev-tools/scripts/StageArtifacts.java:268]: do we need retries?

File: dev-tools/scripts/StageArtifacts.java [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+      var nexus = new NexusApi(params);
+
+      // Collect all files to be uploaded.
+      List<Path> artifacts;
+      try (Stream<Path> pathStream = Files.walk(params.mavenDir)) {
+        artifacts = pathStream
+            .filter(Files::isRegularFile)
+            // Ignore locally generated maven metadata files.
+            .filter(path -> !path.getFileName().toString().startsWith("maven-metadata."))
+            .sorted(Comparator.comparing(Path::toString))
```
Comment (madrob) [inline dev-tools/scripts/StageArtifacts.java:353]: why? do we need parent projects first or something like that?

File: dev-tools/scripts/StageArtifacts.java (hunk 1 above)
Comment (dweiss) [reply dev-tools/scripts/StageArtifacts.java:86]: there's a catch block that does it, below. it's a waste of time to elaborate much on arg validation if it's used from the wizard anyway?

File: dev-tools/scripts/StageArtifacts.java (hunk 3 above)
Comment (dweiss) [reply dev-tools/scripts/StageArtifacts.java:268]: I don't think so. if something fails, it'll fail badly - nexus is not graceful here...

File: dev-tools/scripts/StageArtifacts.java (hunk 1 above)
Comment (dweiss) [reply dev-tools/scripts/StageArtifacts.java:95]: I added that method, just for you, @madrob :)

**Review by madrob — APPROVED** [review]

**Comment by dweiss** [comment]
@jpountz What did you point the script at, exactly? I'll look into it - drop the staging repo, please.

**Comment by dweiss** [comment]
I think I know what the problem is, but would like a confirmation from you first - let me know what you pointed the script at (which directory).

**Comment by dweiss** [comment]
The script should be pointed at the maven folder, including parent subfolders, as in:

https://dist.apache.org/repos/dist/dev/lucene/lucene-9.4.0-RC2-rev-0384b4fcad7856ddc574c8b994c814a568ce6d0a/lucene/maven/

I'm surprised nexus even closed the staging repository since artifact group IDs shouldn't allow you to do that...

**Comment by jpountz** [comment]
This is the folder I ran it against. I used the same folder that the release wizard recommended using with the Ant build.

**Comment by dweiss** [comment]
Give me a sec. you're getting a different output than I do then.

**Comment by dweiss** [comment]
Hi @jpountz . Sorry about this - I'm not sure what happened there but the relative upload path of those artifacts was indeed wrong (and I'm also not sure why nexus isn't complaining about it!...). I fixed the script on 9x and main and added an extra check to make sure the groupId matches every uploaded artifact's path.

You can also add {{--description "Lucene 9x"}} to the command line - it shows as the staging repository's description in the UI.

**Comment by jpountz** [comment]
No worries at all @dweiss, thanks a lot for modernizing Maven staging upload, it's a relief we no longer have a dependency on a 8.x build!

**Comment by dweiss** [comment]
I wonder if this should be reported to infra though. The fact you can close an upload repository with top-level artifacts is strange... I wonder what would happen if you clicked release on that, would it actually promote those JARs to here? It seems crazy!
https://repo1.maven.org/maven2/

**Comment by jpountz** [comment]
@dweiss So I did release the staged repository with top-level artifacts, and Nexus did not give me any errors either. :( After browsing the content, I saw all the artifacts that I had expected, it's only after clicking and realizing that publication to Central wasn't working that I thought that the lack of hierarchy was weird.

Good news is that I have released with the correct directory layout earlier today and now I can see the artifacts on Maven Central.

**Comment by dweiss** [comment]
I filed this question to infra: https://issues.apache.org/jira/browse/INFRA-23931

**Comment by jpountz** [comment]
Thanks Dawid!

**Comment by dweiss** [comment]
Just FYI - Sonatype's nexus wouldn't let you release that staging repository, I've tried:
image

**Comment by dweiss** [comment]
I'll follow up with Gavin on the Infra ticket. Seems like the rules at Apache are not as strict. And sorry for this.

**Comment by jpountz** [comment]
Again, no worries at all, not your fault.

**Pull request merged**